### PR TITLE
Simplify CL API

### PR DESCRIPTION
**Pull request by b-chu**
Changes the API to

- specify the dataset and use the dataloader settings from train_loader
- not need to respecify the first dataset in the callback

Same loss curve as the old callback
image

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Deferring to Daniel

Should @codestar12 also review

File: llmfoundry/callbacks/curriculum_learning_callback.py [hunk 1]
```diff
@@ -53,48 +55,47 @@ class CurriculumLearning(CallbackWithConfig):
...
         schedule: list[dict[str, Any]],
     ):
         # Ensure all duration units are in epochs or tokens and values are positive
         self._schedule = schedule
         if len(self._schedule) == 0:
             raise ValueError('The schedule must have at least one datamix.')
-        for index, datamix in enumerate(self._schedule):
+        first_datamix = {
+            'duration': duration,
+            'dataset': train_config['train_loader']['dataset'],
```
Comment (snarayan21) [inline llmfoundry/callbacks/curriculum_learning_callback.py:82]: question -- did we want to modify how this CL callback is being used or bring back the previous one as well @codestar12 ?

File: llmfoundry/callbacks/curriculum_learning_callback.py [hunk 2]
```diff
@@ -211,18 +213,20 @@ def _build_train_loader(
...
             f'Building train loader in CurriculumLearning callback for dataset {self._schedule_index}',
         )
-        assert self.tokenizer is not None
+        assert self._tokenizer is not None
         try:
             return build_dataloader(
                 train_loader_config,
-                self.tokenizer,
-                self.device_train_batch_size,
+                self._tokenizer,
```
Comment (snarayan21) [inline llmfoundry/callbacks/curriculum_learning_callback.py:228]: just want to make sure, this should work with finetuning dataloader as well right?

**Review by snarayan21 — COMMENTED** [review]
mind doing a before/after manual test that shows that the loss curves are the same with this change?

**Comment by mvpatel2000** [comment]
We should ensure @codestar12 reviews

File: llmfoundry/callbacks/curriculum_learning_callback.py [hunk 3]
```diff
@@ -53,48 +55,47 @@ class CurriculumLearning(CallbackWithConfig):
...
-            ):
-                raise ValueError((
-                    'The first datamix in the schedule must match the '
-                    'train_loader in the train_config.'
-                ))
-
         self._schedule_index = 0
-        self.device_train_batch_size = train_config['device_train_batch_size']
-        self.tokenizer = None
+        self._train_loader_config: dict[str, Any] = train_config['train_loader']
```
[the remark this answers is not in the document]
Comment (b-chu) [reply llmfoundry/callbacks/curriculum_learning_callback.py:89]: Yes

File: llmfoundry/callbacks/curriculum_learning_callback.py (hunk 2 above)
Comment (b-chu) [reply llmfoundry/callbacks/curriculum_learning_callback.py:228]: Yes

File: llmfoundry/callbacks/curriculum_learning_callback.py (hunk 1 above)
Comment (snarayan21) [reply llmfoundry/callbacks/curriculum_learning_callback.py:82]: what was the decision lol

File: llmfoundry/callbacks/curriculum_learning_callback.py [hunk 4]
```diff
@@ -53,48 +57,59 @@ class CurriculumLearning(CallbackWithConfig):
...
+        duration: Optional[Union[Time, str, int]] = None,
     ):
+        if duration is None:
+            warnings.warn(
+                VersionedDeprecationWarning(
+                    'Specifying the full schedule in the CurriculumLearning ' +
+                    'callback is deprecated. Please specify the duration of ' +
+                    'the first datamix separately and change the schedule ' +
+                    'use datasets instead of dataloaders.',
+                    remove_version='0.13.0'
```
Comment (dakinggg) [inline llmfoundry/callbacks/curriculum_learning_callback.py:85]: should be 0.15.0, right? So the deprecation warning is included in two releases (13 and 14)

**Review by dakinggg — APPROVED** [review]

File: llmfoundry/callbacks/curriculum_learning_callback.py (hunk 4 above)
Comment (b-chu) [reply llmfoundry/callbacks/curriculum_learning_callback.py:85]: Done
